spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/redis/RedisTokenStore.java

```diff
@@ -398,24 +399,68 @@ private void removeAccessTokenUsingRefreshToken(String refreshToken) {
 		}
 	}
 
-	private List<byte[]> getByteLists(byte[] approvalKey, RedisConnection conn) {
+	private List<byte[]> getZByteLists(byte[] key, RedisConnection conn) {
+		// Sorted Set expiration maintenance
+		long currentTime = System.currentTimeMillis();
+		conn.zRemRangeByScore(key, 0, currentTime);
+
 		List<byte[]> byteList;
-		Long size = conn.sCard(approvalKey);
+		Long size = conn.zCard(key);
 		byteList = new ArrayList<byte[]>(size.intValue());
-		Cursor<byte[]> cursor = conn.sScan(approvalKey, ScanOptions.NONE);
+		Cursor<RedisZSetCommands.Tuple> cursor = conn.zScan(key, ScanOptions.NONE);
+
 		while(cursor.hasNext()) {
-			byteList.add(cursor.next());
+			RedisZSetCommands.Tuple t = cursor.next();
+
+			// Probably not necessary because of the maintenance at the beginning but why not...
+			if (t.getScore() == -1 || t.getScore() > currentTime) {
+				byteList.add(t.getValue());
+			}
 		}
 		return byteList;
 	}
 
+	/**
+	 * Runs a maintenance of the RedisTokenStore.
+	 *
+	 * SortedSets UNAME_TO_ACCESS and CLIENT_ID_TO_ACCESS contains access tokens that can expire.
+	 * This expiration is set as a score of the Redis SortedSet data structure. Redis does not support expiration of items in a container data structure.
+	 * It supports only expiration of whole key. In case there is still new access tokens being stored into the RedisTokenStore before whole key gets expired,
+	 * the expiration is prolonged and the key is not effectively deleted. To do "garbage collection" this method should be called once upon a time.
+	 * @return how many items were removed
+	 */
+	public long doMaintenance() {
```

Comment:
This cannot go into a patch release since it's `public`. Please see the [changes allowed](https://apr.apache.org/versioning.html#strategy) for a patch release.

Also, I don't think it makes sense to have a "maintenance" routine in the `TokenStore` implementation. I think this responsibility should be in a separate component altogether.

Reply:
How to maintain the token store depends on the internal details of the token store. I'm not sure a separate component would be a good design.